### ISSUE TITLE
mmCIF parser and insertion codes

### ISSUE THREAD

**Issue by ingvarlagerstedt** [biopython/biopython#481]
I get an exception when attempting to parse an mmCIF file with insertion codes, the corresponding PDB file parses fine.

When using auth_seq_id, then pdbx_PDB_ins_code should be used for the insertion code. When using label_seq_id, there is no need to use pdbx_PDB_ins_code.

I used 3v11 as an example, which throws an exception on atom 5662, the first atom in the second residue 17 in chain D.

The easy fix is to just ignore auth_seq_id.  The slightly harder is to handle pdbx_PDB_ins_code.

My example program:

```
#!/usr/bin/python 
import sys
from traceback import print_exc
from Bio.PDB import MMCIFParser
from Bio.PDB import PDBParser

def proc(eId):
  structure_id = eId.upper()
  filename = eId+".cif"
  try:
    p=MMCIFParser()
    structure=p.get_structure(structure_id, filename)
  except:
    print "Exception"
    print_exc()

  filename = eId+".pdb"
  try:
    pdb = PDBParser()
    structure=pdb.get_structure(structure_id, filename)
  except:
    print "Exception"
    print_exc()

def main(argv):

  for e in ['3v11',]:
    proc(e)

if __name__ == "__main__":
    main(sys.argv[1:])
```

gave the following output:

```
Exception
Traceback (most recent call last):
  File "mmCIF.py", line 12, in proc
    structure=p.get_structure(structure_id, filename)
  File "/Library/Python/2.7/site-packages/biopython-1.65-py2.7-macosx-10.8-intel.egg/Bio/PDB/MMCIFParser.py", line 25, in get_structure
    self._build_structure(structure_id)
  File "/Library/Python/2.7/site-packages/biopython-1.65-py2.7-macosx-10.8-intel.egg/Bio/PDB/MMCIFParser.py", line 131, in _build_structure
    name, element=element)
  File "/Library/Python/2.7/site-packages/biopython-1.65-py2.7-macosx-10.8-intel.egg/Bio/PDB/StructureBuilder.py", line 234, in init_atom
    residue.add(atom)
  File "/Library/Python/2.7/site-packages/biopython-1.65-py2.7-macosx-10.8-intel.egg/Bio/PDB/Residue.py", line 82, in add
    "Atom %s defined twice in residue %s" % (atom_id, self))
PDBConstructionException: Atom P defined twice in residue <Residue C het=  resseq=17 icode= >
/Library/Python/2.7/site-packages/biopython-1.65-py2.7-macosx-10.8-intel.egg/Bio/PDB/StructureBuilder.py:86: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7784.
  PDBConstructionWarning)
```

Kind Regards,
Ingvar Lagerstedt
PDBe, EMDB-EBI

**Comment by ingvarlagerstedt** [comment]
The easy fix suggested does not work,  mmCIF only have . for label_seq_id.  So switching to label_seq_id would fail on any entry with a ligand/ion and I think water.
To use auth_seq_id, the following is needed:
1. fetch insertion codes
icode_list = mmcif_dict["_atom_site.pdbx_PDB_ins_code"]
2. rewrite _get_icode() to something like - this is wrong/not needed for label_seq_id 
def _get_icode(self, resseq, icode):
     """ This only works for auth_seq_id.  Ligands do not have sequence numbers
     label_seq_id. The icode is in pdbx_PDB_ins_code, and is only used together
    with auth_seq_id """
    if icode == '?': icode = ""

    return icode, int(resseq)
3. _get_icode need to be check for each atom, as resseq does not change for insertion code residues when using auth_seq_id.  icode needs to be tracked with current_icode, in the same way as resseq is.  Something like:      
    icode, int_resseq = self._get_icode(resseq, icode_list[i])
    if current_chain_id != chainid:
       current_chain_id = chainid
       structure_builder.init_chain(current_chain_id)
       current_residue_id = resseq
       current_icode = icode
       structure_builder.init_residue(resname, hetatm_flag, int_resseq, icode)
    elif current_residue_id != resseq or current_icode != icode:
      current_residue_id = resseq
      current_icode = icode
      structure_builder.init_residue(resname, hetatm_flag, int_resseq, icode)

**Comment by peterjc** [comment]
You need to use the triple back-tick character to mark blocks of code in GitHub comments (I've edited your post), or double back-tick for variable names (less important).

**Comment by peterjc** [comment]
@lennax - could you look at this issue?

**Comment by biomadeira** [comment]
My last PR on the MMCIF parser https://github.com/biopython/biopython/pull/502 should fix this problem as the insertion code is coming from pdbx_PDB_ins_code instead of being parsed from the sequence (as was previously defined in method shown 3 comments above). I this particular structure though I think the problem is actually in the mmCIF not in biopython. As you can see below Atom P is defined twice for the auth_seq_id 17. In fact in this case if the instead of auth_seq_id we were using the label_seq_id this wouldn't be a problem as it changes to 18. The problem I guess is that it isn't uniform so using either auth_seq_id or label_seq_id isn't failure free for all cases. 

loop_
_atom_site.group_PDB 
_atom_site.id 
_atom_site.type_symbol 
_atom_site.label_atom_id 
_atom_site.label_alt_id 
_atom_site.label_comp_id 
_atom_site.label_asym_id 
_atom_site.label_entity_id 
_atom_site.label_seq_id 
_atom_site.pdbx_PDB_ins_code 
_atom_site.Cartn_x 
_atom_site.Cartn_y 
_atom_site.Cartn_z 
_atom_site.occupancy 
_atom_site.B_iso_or_equiv 
_atom_site.Cartn_x_esd 
_atom_site.Cartn_y_esd 
_atom_site.Cartn_z_esd 
_atom_site.occupancy_esd 
_atom_site.B_iso_or_equiv_esd 
_atom_site.pdbx_formal_charge 
_atom_site.auth_seq_id 
_atom_site.auth_comp_id 
_atom_site.auth_asym_id 
_atom_site.auth_atom_id 
_atom_site.pdbx_PDB_model_num  
ATOM   5642 P  P     . C   D 4 17  ? 67.059  36.163 -1.605  1.00 300.00 ? ? ? ? ? ? 17  C   D P     1 
ATOM   5643 O  OP1   . C   D 4 17  ? 66.432  36.138 -2.953  1.00 300.00 ? ? ? ? ? ? 17  C   D OP1   1 
ATOM   5644 O  OP2   . C   D 4 17  ? 68.428  35.611 -1.449  1.00 300.00 ? ? ? ? ? ? 17  C   D OP2   1 
ATOM   5645 O  "O5'" . C   D 4 17  ? 67.005  37.633 -1.016  1.00 300.00 ? ? ? ? ? ? 17  C   D "O5'" 1 
ATOM   5646 C  "C5'" . C   D 4 17  ? 66.022  37.992 -0.067  1.00 300.00 ? ? ? ? ? ? 17  C   D "C5'" 1 
ATOM   5647 C  "C4'" . C   D 4 17  ? 66.538  39.018 0.895   1.00 300.00 ? ? ? ? ? ? 17  C   D "C4'" 1 
ATOM   5648 O  "O4'" . C   D 4 17  ? 67.253  38.367 1.976   1.00 300.00 ? ? ? ? ? ? 17  C   D "O4'" 1 
ATOM   5649 C  "C3'" . C   D 4 17  ? 67.556  39.990 0.336   1.00 300.00 ? ? ? ? ? ? 17  C   D "C3'" 1 
ATOM   5650 O  "O3'" . C   D 4 17  ? 66.987  41.022 -0.433  1.00 300.00 ? ? ? ? ? ? 17  C   D "O3'" 1 
ATOM   5651 C  "C2'" . C   D 4 17  ? 68.256  40.486 1.584   1.00 300.00 ? ? ? ? ? ? 17  C   D "C2'" 1 
ATOM   5652 O  "O2'" . C   D 4 17  ? 67.451  41.438 2.264   1.00 300.00 ? ? ? ? ? ? 17  C   D "O2'" 1 
ATOM   5653 C  "C1'" . C   D 4 17  ? 68.308  39.206 2.414   1.00 300.00 ? ? ? ? ? ? 17  C   D "C1'" 1 
ATOM   5654 N  N1    . C   D 4 17  ? 69.608  38.504 2.246   1.00 300.00 ? ? ? ? ? ? 17  C   D N1    1 
ATOM   5655 C  C2    . C   D 4 17  ? 69.833  37.348 1.486   1.00 300.00 ? ? ? ? ? ? 17  C   D C2    1 
ATOM   5656 O  O2    . C   D 4 17  ? 68.910  36.780 0.903   1.00 300.00 ? ? ? ? ? ? 17  C   D O2    1 
ATOM   5657 N  N3    . C   D 4 17  ? 71.083  36.836 1.412   1.00 300.00 ? ? ? ? ? ? 17  C   D N3    1 
ATOM   5658 C  C4    . C   D 4 17  ? 72.098  37.417 2.035   1.00 300.00 ? ? ? ? ? ? 17  C   D C4    1 
ATOM   5659 N  N4    . C   D 4 17  ? 73.301  36.865 1.907   1.00 300.00 ? ? ? ? ? ? 17  C   D N4    1 
ATOM   5660 C  C5    . C   D 4 17  ? 71.917  38.601 2.792   1.00 300.00 ? ? ? ? ? ? 17  C   D C5    1 
ATOM   5661 C  C6    . C   D 4 17  ? 70.677  39.093 2.851   1.00 300.00 ? ? ? ? ? ? 17  C   D C6    1 
ATOM   5662 P  P     . U   D 4 18  A 67.874  41.732 -1.556  1.00 300.00 ? ? ? ? ? ? 17  U   D P     1 

Later in that same structure there are some hetatms with no valid label_seq_id 
HETATM 6935 N  N     . MET E 5 .   ? 19.037  48.175 14.227  1.00 300.00 ? ? ? ? ? ? 101 MET D N     1 
HETATM 6936 C  CA    . MET E 5 .   ? 19.677  47.219 15.167  1.00 300.00 ? ? ? ? ? ? 101 MET D CA    1 
HETATM 6937 C  C     . MET E 5 .   ? 20.938  46.560 14.657  1.00 300.00 ? ? ? ? ? ? 101 MET D C     1 
HETATM 6938 O  O     . MET E 5 .   ? 20.973  45.620 13.868  1.00 300.00 ? ? ? ? ? ? 101 MET D O     1 
HETATM 6939 C  CB    . MET E 5 .   ? 18.844  46.281 15.975  1.00 300.00 ? ? ? ? ? ? 101 MET D CB    1 
HETATM 6940 C  CG    . MET E 5 .   ? 19.142  46.337 17.464  1.00 300.00 ? ? ? ? ? ? 101 MET D CG    1 
HETATM 6941 S  SD    . MET E 5 .   ? 20.927  46.263 17.807  1.00 300.00 ? ? ? ? ? ? 101 MET D SD    1 
HETATM 6942 C  CE    . MET E 5 .   ? 21.405  44.604 17.936  1.00 300.00 ? ? ? ? ? ? 101 MET D CE    1 
HETATM 6943 MG MG    . MG  F 6 .   ? 23.821  31.899 25.521  1.00 300.00 ? ? ? ? ? ? 501 MG  A MG    1 

I don't know what would be a solid fix. At the moment it parses auth_seq_id if available otherwise label_seq_id (https://github.com/biopython/biopython/blob/master/Bio/PDB/MMCIFParser.py#L99). Also I just realized how dangerous this int(resseq) is https://github.com/biopython/biopython/blob/master/Bio/PDB/MMCIFParser.py#L167 ...

**Comment by ingvarlagerstedt** [comment]
Dear Fabio,

I am little bit confused about the statement that the mmCIF file is the problem here.  Non-polymer residues/ligands do not have a residue number in mmCIF, while they do in PDB.  And, I thought the whole point of insertion codes was that the auth_seq_id could be the same for two residues.  
Take an entry like 1afe, there are many inserted residues at the top and a few ligands at the end.  Are you saying that this (any many other) distributed mmCIF are wrong?

Kind Regards,
Ingvar Lagerstedt

loop_
_atom_site.group_PDB 
_atom_site.id 
_atom_site.type_symbol 
_atom_site.label_atom_id 
_atom_site.label_alt_id 
_atom_site.label_comp_id 
_atom_site.label_asym_id 
_atom_site.label_entity_id 
_atom_site.label_seq_id 
_atom_site.pdbx_PDB_ins_code 
_atom_site.Cartn_x 
_atom_site.Cartn_y 
_atom_site.Cartn_z 
_atom_site.occupancy 
_atom_site.B_iso_or_equiv 
_atom_site.Cartn_x_esd 
_atom_site.Cartn_y_esd 
_atom_site.Cartn_z_esd 
_atom_site.occupancy_esd 
_atom_site.B_iso_or_equiv_esd 
_atom_site.pdbx_formal_charge 
_atom_site.auth_seq_id 
_atom_site.auth_comp_id 
_atom_site.auth_asym_id 
_atom_site.auth_atom_id 
_atom_site.pdbx_PDB_model_num 
ATOM   1    N N   . THR A 1 1   H 18.295 13.806  18.609  0.00 93.50  ? ? ? ? ? ? 1   THR L N   1 
ATOM   2    C CA  . THR A 1 1   H 19.671 13.519  18.253  0.00 93.66  ? ? ? ? ? ? 1   THR L CA  1 
ATOM   3    C C   . THR A 1 1   H 20.531 13.236  19.471  0.00 50.00  ? ? ? ? ? ? 1   THR L C   1 
ATOM   4    O O   . THR A 1 1   H 21.627 12.705  19.357  0.00 50.00  ? ? ? ? ? ? 1   THR L O   1 
ATOM   5    C CB  . THR A 1 1   H 20.324 14.689  17.472  0.00 99.48  ? ? ? ? ? ? 1   THR L CB  1 
ATOM   6    O OG1 . THR A 1 1   H 19.670 14.926  16.246  0.00 98.62  ? ? ? ? ? ? 1   THR L OG1 1 
ATOM   7    C CG2 . THR A 1 1   H 21.801 14.401  17.233  0.00 97.64  ? ? ? ? ? ? 1   THR L CG2 1 
ATOM   8    N N   . PHE A 1 2   G 20.073 13.650  20.644  0.00 96.42  ? ? ? ? ? ? 1   PHE L N   1 
ATOM   9    C CA  . PHE A 1 2   G 20.897 13.530  21.818  0.00 92.46  ? ? ? ? ? ? 1   PHE L CA  1 
ATOM   10   C C   . PHE A 1 2   G 20.110 13.497  23.107  0.00 50.00  ? ? ? ? ? ? 1   PHE L C   1 
ATOM   11   O O   . PHE A 1 2   G 19.128 12.778  23.244  0.00 59.28  ? ? ? ? ? ? 1   PHE L O   1 
ATOM   12   C CB  . PHE A 1 2   G 21.870 14.723  21.862  0.00 91.34  ? ? ? ? ? ? 1   PHE L CB  1 
ATOM   13   C CG  . PHE A 1 2   G 21.129 16.046  21.882  0.00 89.99  ? ? ? ? ? ? 1   PHE L CG  1 
ATOM   14   C CD1 . PHE A 1 2   G 20.262 16.392  20.843  0.00 89.22  ? ? ? ? ? ? 1   PHE L CD1 1 
ATOM   15   C CD2 . PHE A 1 2   G 21.218 16.901  22.980  0.00 88.06  ? ? ? ? ? ? 1   PHE L CD2 1 
ATOM   16   C CE1 . PHE A 1 2   G 19.545 17.588  20.863  0.00 88.40  ? ? ? ? ? ? 1   PHE L CE1 1 
ATOM   17   C CE2 . PHE A 1 2   G 20.510 18.102  23.019  0.00 88.95  ? ? ? ? ? ? 1   PHE L CE2 1 
ATOM   18   C CZ  . PHE A 1 2   G 19.672 18.443  21.957  0.00 86.51  ? ? ? ? ? ? 1   PHE L CZ  1 
ATOM   19   N N   . GLY A 1 3   F 20.609 14.276  24.083  0.00 50.00  ? ? ? ? ? ? 1   GLY L N   1 
ATOM   20   C CA  . GLY A 1 3   F 19.991 14.387  25.414  0.00 50.00  ? ? ? ? ? ? 1   GLY L CA  1 
ATOM   21   C C   . GLY A 1 3   F 18.528 14.826  25.308  0.00 50.00  ? ? ? ? ? ? 1   GLY L C   1 
ATOM   22   O O   . GLY A 1 3   F 18.194 15.649  24.454  0.00 99.53  ? ? ? ? ? ? 1   GLY L O   1 
ATOM   23   N N   . SER A 1 4   E 17.627 14.253  26.150  1.00 89.35  ? ? ? ? ? ? 1   SER L N   1 
ATOM   24   C CA  . SER A 1 4   E 16.203 14.587  26.050  1.00 79.94  ? ? ? ? ? ? 1   SER L CA  1 
ATOM   25   C C   . SER A 1 4   E 15.775 14.281  24.659  0.00 62.01  ? ? ? ? ? ? 1   SER L C   1 
ATOM   26   O O   . SER A 1 4   E 14.646 14.521  24.247  0.00 61.63  ? ? ? ? ? ? 1   SER L O   1 
ATOM   27   C CB  . SER A 1 4   E 15.926 16.032  26.412  0.00 83.73  ? ? ? ? ? ? 1   SER L CB  1 
ATOM   28   O OG  . SER A 1 4   E 14.578 16.176  26.811  0.00 92.60  ? ? ? ? ? ? 1   SER L OG  1 
ATOM   29   N N   . GLY A 1 5   D 16.745 13.745  23.945  0.00 40.53  ? ? ? ? ? ? 1   GLY L N   1 
ATOM   30   C CA  . GLY A 1 5   D 16.612 13.330  22.603  1.00 31.42  ? ? ? ? ? ? 1   GLY L CA  1 
ATOM   31   C C   . GLY A 1 5   D 16.779 14.435  21.568  0.00 42.89  ? ? ? ? ? ? 1   GLY L C   1 
ATOM   32   O O   . GLY A 1 5   D 17.893 14.773  21.174  0.00 42.66  ? ? ? ? ? ? 1   GLY L O   1 
ATOM   33   N N   . GLU A 1 6   C 15.657 14.961  21.101  0.00 45.10  ? ? ? ? ? ? 1   GLU L N   1 
ATOM   34   C CA  . GLU A 1 6   C 15.647 15.972  20.062  0.00 47.36  ? ? ? ? ? ? 1   GLU L CA  1 
ATOM   35   C C   . GLU A 1 6   C 15.254 17.361  20.583  0.00 53.44  ? ? ? ? ? ? 1   GLU L C   1 
ATOM   36   O O   . GLU A 1 6   C 15.675 17.781  21.660  0.00 52.85  ? ? ? ? ? ? 1   GLU L O   1 
ATOM   37   C CB  . GLU A 1 6   C 14.700 15.517  18.944  0.00 49.29  ? ? ? ? ? ? 1   GLU L CB  1 
ATOM   38   C CG  . GLU A 1 6   C 15.346 15.470  17.572  0.00 63.92  ? ? ? ? ? ? 1   GLU L CG  1 
ATOM   39   C CD  . GLU A 1 6   C 15.059 16.724  16.847  0.00 58.25  ? ? ? ? ? ? 1   GLU L CD  1 
ATOM   40   O OE1 . GLU A 1 6   C 14.278 17.554  17.407  0.00 54.21  ? ? ? ? ? ? 1   GLU L OE1 1 
ATOM   41   O OE2 . GLU A 1 6   C 15.646 16.926  15.754  0.00 50.00  ? ? ? ? ? ? 1   GLU L OE2 1 
ATOM   42   N N   . ALA A 1 7   B 14.441 18.060  19.784  1.00 52.33  ? ? ? ? ? ? 1   ALA L N   1 
ATOM   43   C CA  . ALA A 1 7   B 13.884 19.410  20.100  1.00 56.86  ? ? ? ? ? ? 1   ALA L CA  1 
ATOM   44   C C   . ALA A 1 7   B 12.363 19.411  20.024  1.00 52.57  ? ? ? ? ? ? 1   ALA L C   1 
ATOM   45   O O   . ALA A 1 7   B 11.652 19.951  20.873  1.00 44.02  ? ? ? ? ? ? 1   ALA L O   1 
ATOM   46   C CB  . ALA A 1 7   B 14.410 20.462  19.156  1.00 58.72  ? ? ? ? ? ? 1   ALA L CB  1 
ATOM   47   N N   . ASP A 1 8   A 11.903 18.837  18.956  1.00 46.33  ? ? ? ? ? ? 1   ASP L N   1 
ATOM   48   C CA  . ASP A 1 8   A 10.534 18.582  18.735  1.00 44.94  ? ? ? ? ? ? 1   ASP L CA  1 
ATOM   49   C C   . ASP A 1 8   A 10.249 17.117  19.130  1.00 36.45  ? ? ? ? ? ? 1   ASP L C   1 
ATOM   50   O O   . ASP A 1 8   A 9.142  16.570  18.965  1.00 25.84  ? ? ? ? ? ? 1   ASP L O   1 
ATOM   51   C CB  . ASP A 1 8   A 10.282 18.711  17.249  1.00 49.75  ? ? ? ? ? ? 1   ASP L CB  1 
ATOM   52   C CG  . ASP A 1 8   A 8.989  19.336  17.094  1.00 81.13  ? ? ? ? ? ? 1   ASP L CG  1 
ATOM   53   O OD1 . ASP A 1 8   A 8.462  19.759  18.181  1.00 72.62  ? ? ? ? ? ? 1   ASP L OD1 1 
ATOM   54   O OD2 . ASP A 1 8   A 8.537  19.511  15.912  1.00 100.00 ? ? ? ? ? ? 1   ASP L OD2 1 
ATOM   55   N N   . CYS A 1 9   ? 11.315 16.455  19.582  1.00 24.10  ? ? ? ? ? ? 1   CYS L N   1 
ATOM   56   C CA  . CYS A 1 9   ? 11.235 15.054  19.876  1.00 23.85  ? ? ? ? ? ? 1   CYS L CA  1 
ATOM   57   C C   . CYS A 1 9   ? 10.080 14.643  20.739  1.00 21.24  ? ? ? ? ? ? 1   CYS L C   1 
ATOM   58   O O   . CYS A 1 9   ? 9.627  15.355  21.640  1.00 24.64  ? ? ? ? ? ? 1   CYS L O   1 
ATOM   59   C CB  . CYS A 1 9   ? 12.586 14.465  20.388  1.00 26.98  ? ? ? ? ? ? 1   CYS L CB  1 
ATOM   60   S SG  . CYS A 1 9   ? 12.985 14.961  22.146  1.00 29.26  ? ? ? ? ? ? 1   CYS L SG  1 
ATOM   61   N N   . GLY A 1 10  ? 9.652  13.410  20.492  1.00 20.70  ? ? ? ? ? ? 2   GLY L N   1 
ATOM   62   C CA  . GLY A 1 10  ? 8.648  12.726  21.313  1.00 19.18  ? ? ? ? ? ? 2   GLY L CA  1 
ATOM   63   C C   . GLY A 1 10  ? 7.259  13.304  21.310  1.00 21.72  ? ? ? ? ? ? 2   GLY L C   1 
ATOM   64   O O   . GLY A 1 10  ? 6.429  12.833  22.081  1.00 24.26  ? ? ? ? ? ? 2   GLY L O   1 
...
HETATM 2436 C C1  . NAG D 4 .   ? 31.987 -8.260  10.862  0.00 55.00  ? ? ? ? ? ? 400 NAG H C1  1 
HETATM 2437 C C2  . NAG D 4 .   ? 33.051 -8.574  11.874  0.00 55.00  ? ? ? ? ? ? 400 NAG H C2  1 
HETATM 2438 C C3  . NAG D 4 .   ? 34.425 -7.954  11.704  0.00 55.00  ? ? ? ? ? ? 400 NAG H C3  1 
HETATM 2439 C C4  . NAG D 4 .   ? 34.644 -7.408  10.319  0.00 55.00  ? ? ? ? ? ? 400 NAG H C4  1 
HETATM 2440 C C5  . NAG D 4 .   ? 33.460 -7.223  9.342   0.00 55.00  ? ? ? ? ? ? 400 NAG H C5  1 
HETATM 2441 C C6  . NAG D 4 .   ? 33.758 -6.574  7.993   0.00 55.00  ? ? ? ? ? ? 400 NAG H C6  1 
HETATM 2442 C C7  . NAG D 4 .   ? 33.411 -10.523 13.568  0.00 55.00  ? ? ? ? ? ? 400 NAG H C7  1 
HETATM 2443 C C8  . NAG D 4 .   ? 33.428 -11.772 12.709  0.00 55.00  ? ? ? ? ? ? 400 NAG H C8  1 
HETATM 2444 N N2  . NAG D 4 .   ? 32.686 -9.488  13.012  0.00 55.00  ? ? ? ? ? ? 400 NAG H N2  1 
HETATM 2445 O O3  . NAG D 4 .   ? 35.591 -8.494  12.375  0.00 55.00  ? ? ? ? ? ? 400 NAG H O3  1 
HETATM 2446 O O4  . NAG D 4 .   ? 35.934 -6.790  9.994   0.00 55.00  ? ? ? ? ? ? 400 NAG H O4  1 
HETATM 2447 O O5  . NAG D 4 .   ? 32.424 -8.192  9.495   0.00 55.00  ? ? ? ? ? ? 400 NAG H O5  1 
HETATM 2448 O O6  . NAG D 4 .   ? 32.772 -6.344  7.026   0.00 55.00  ? ? ? ? ? ? 400 NAG H O6  1 
HETATM 2449 O O7  . NAG D 4 .   ? 33.006 -10.762 14.706  0.00 55.00  ? ? ? ? ? ? 400 NAG H O7  1

**Comment by biomadeira** [comment]
@ingvarlagerstedt 
Dear Ingvar,
Thanks for your comment. I guess you are right! I was a bit confused with these label_seq_id and auth_seq_id fields in the mmCIF. I have made another PR https://github.com/biopython/biopython/pull/523 which should fix the problem for good this time.
Thanks, F.

**Comment by peterjc** [comment]
We closed #523, and I just checked the original example from @ingvarlagerstedt  for this issue and that passes now - also on Python 3 if updated to use print functions:

```
$ python3 i481.py 
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6934.
  PDBConstructionWarning)
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 6967.
  PDBConstructionWarning)
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7784.
  PDBConstructionWarning)
```

So, while I've not worked out when it was fixed, it works now. Thanks everyone - closing this.
